File: apps/server/src/modules/loyalty/loyalty.service.ts
import { prisma } from '@delivery/database';
import { ApiError } from '../../utils/ApiError.js';
import { randomBytes } from 'node:crypto';

const POINTS_PER_DELIVERY = 10;
const REFERRAL_BONUS = 50;

export class LoyaltyService {
  async getProfile(userId: string) {
    const customer = await prisma.customer.findUnique({ where: { userId } });
    if (!customer) throw ApiError.badRequest('Customer profile not found');

    let referralCode = customer.referralCode;
    if (!referralCode) {
      referralCode = `GZ${randomBytes(3).toString('hex').toUpperCase()}`;
      await prisma.customer.update({ where: { id: customer.id }, data: { referralCode } });
    }

    return {
      loyaltyPoints: customer.loyaltyPoints,
      referralCode,
      pointsPerDelivery: POINTS_PER_DELIVERY,
      referralBonus: REFERRAL_BONUS,
    };
  }

  async applyReferral(userId: string, code: string) {
    const customer = await prisma.customer.findUnique({ where: { userId } });
    if (!customer) throw ApiError.badRequest('Customer profile not found');
    if (customer.referredById) throw ApiError.badRequest('Referral already applied');

    const referrer = await prisma.customer.findFirst({ where: { referralCode: code.trim().toUpperCase() } });
    if (!referrer) throw ApiError.notFound('Invalid referral code');
    if (referrer.id === customer.id) throw ApiError.badRequest('Cannot refer yourself');

    await prisma.$transaction([
      prisma.customer.update({ where: { id: customer.id }, data: { referredById: referrer.id } }),
      prisma.customer.update({
        where: { id: referrer.id },
        data: { loyaltyPoints: { increment: REFERRAL_BONUS } },
      }),
      prisma.customer.update({
        where: { id: customer.id },
        data: { loyaltyPoints: { increment: REFERRAL_BONUS } },
      }),
    ]);

    return { applied: true, bonus: REFERRAL_BONUS };
  }

  async awardDeliveryPoints(customerId: string) {
    await prisma.customer.update({
      where: { id: customerId },
      data: { loyaltyPoints: { increment: POINTS_PER_DELIVERY } },
    });
  }
}

export const loyaltyService = new LoyaltyService();

File: apps/guzo-api/src/main/java/et/guzo/service/LoyaltyService.java
package et.guzo.service;

import et.guzo.common.ApiException;
import et.guzo.domain.entity.Customer;
import et.guzo.repository.CustomerRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.SecureRandom;
import java.util.HexFormat;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class LoyaltyService {

    private static final SecureRandom RANDOM = new SecureRandom();
    private static final int POINTS_PER_DELIVERY = 10;
    private static final int REFERRAL_BONUS = 50;

    private final CustomerRepository customerRepository;

    @Transactional
    public Map<String, Object> getProfile(String userId) {
        Customer customer = customerRepository.findByUserId(userId)
            .orElseThrow(() -> ApiException.badRequest("Customer profile not found"));
        if (customer.getReferralCode() == null || customer.getReferralCode().isBlank()) {
            customer.setReferralCode("GZ" + HexFormat.of().formatHex(randomBytes(3)).toUpperCase());
            customerRepository.save(customer);
        }
        return Map.of(
            "loyaltyPoints", customer.getLoyaltyPoints(),
            "referralCode", customer.getReferralCode(),
            "pointsPerDelivery", POINTS_PER_DELIVERY,
            "referralBonus", REFERRAL_BONUS
        );
    }

    @Transactional
    public Map<String, Object> applyReferral(String userId, String code) {
        Customer customer = customerRepository.findByUserId(userId)
            .orElseThrow(() -> ApiException.badRequest("Customer profile not found"));
        if (customer.getReferredById() != null) {
            throw ApiException.badRequest("Referral already applied");
        }
        Customer referrer = customerRepository.findByReferralCodeIgnoreCase(code.trim())
            .orElseThrow(() -> ApiException.notFound("Invalid referral code"));
        if (referrer.getId().equals(customer.getId())) {
            throw ApiException.badRequest("Cannot refer yourself");
        }
        customer.setReferredById(referrer.getId());
        customer.setLoyaltyPoints(customer.getLoyaltyPoints() + REFERRAL_BONUS);
        referrer.setLoyaltyPoints(referrer.getLoyaltyPoints() + REFERRAL_BONUS);
        customerRepository.save(customer);
        customerRepository.save(referrer);
        return Map.of("applied", true, "bonus", REFERRAL_BONUS);
    }

    private static byte[] randomBytes(int len) {
        byte[] b = new byte[len];
        RANDOM.nextBytes(b);
        return b;
    }
}
